File: app/src/main/java/android/voguepay/ng/voguepaysdk/Ui/payment/Repository.java
package android.voguepay.ng.voguepaysdk.Ui.payment;

import android.util.Log;
import android.voguepay.ng.voguepaysdk.AppExecutors;
import android.voguepay.ng.voguepaysdk.WhiteLabel;
import android.voguepay.ng.voguepaysdk.api.VoguePayApiService;
import android.voguepay.ng.voguepaysdk2.LabelResponse;
import android.voguepay.ng.voguepaysdk2.Payment;
import androidx.lifecycle.MutableLiveData;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

import javax.inject.Inject;

public class Repository {

    @Inject
    VoguePayApiService  apiService;

    private Disposable searchDisposable;

    String authUrl;



    AppExecutors appExecutors;

    @Inject
    public Repository(AppExecutors appExecutors) {
        this.appExecutors = appExecutors;
    }


    public MutableLiveData<String> getLink (Payment payment){
        if (searchDisposable != null){
            searchDisposable.dispose();
        }

        MutableLiveData<String> url = new MutableLiveData<>();
         apiService.getPaymentUrl(payment.getP(),
                 payment.getV_merchant_id(),
                 payment.getTotal(),
                 payment.getMemo(),
                 payment.getCur())
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<String>() {
                               @Override
                               public void onSubscribe(Disposable d) {
                                   searchDisposable = d;
                               }

                               @Override
                               public void onNext(String s) {
                                   url.setValue(s);

                               }

                               @Override
                               public void onError(Throwable e) {
                                   Log.e("Rep", e.getMessage());

                               }

                               @Override
                               public void onComplete() {
                                   Log.e("TAG", "onComplete was reached");

                               }
                           }

                );

         return  url;


    }
    public MutableLiveData<LabelResponse> getLink (WhiteLabel payment){


        if (searchDisposable != null){
            searchDisposable.dispose();
        }
        Gson gson = new GsonBuilder().create();
        String payload = gson.toJson(payment);
        Log.e("repository", payload);


        MutableLiveData<LabelResponse> url = new MutableLiveData<>();
         apiService.getPaymentHeader(payload)
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<LabelResponse>() {
                               @Override
                               public void onSubscribe(Disposable d) {
                                   searchDisposable = d;
                               }

                               @Override
                               public void onNext(LabelResponse s) {
                                   url.setValue(s);
                                   if (s != null) {
                                       Log.e("Rep", s.toString());
                                   }
                               }

                               @Override
                               public void onError(Throwable e) {
                                   Log.e("Rep", e.getMessage());

                               }

                               @Override
                               public void onComplete() {
                                   Log.e("TAG", "onComplete was reached");

                               }
                           }

                );

         return  url;


    }
}

File: app/src/main/java/android/voguepay/ng/voguepaysdk/ui/payment/Repository.kt
package android.voguepay.ng.voguepaysdk.ui.payment

import android.util.Log
import android.voguepay.ng.voguepaysdk.WhiteLabel
import android.voguepay.ng.voguepaysdk.api.VoguePayApiService
import android.voguepay.ng.voguepaysdk2.LabelResponse
import android.voguepay.ng.voguepaysdk2.Payment
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.google.gson.GsonBuilder
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import retrofit2.Retrofit
import javax.inject.Inject

class Repository @Inject constructor(
    retrofit: Retrofit) {

    private  val apiService   = retrofit.create(VoguePayApiService::class.java)


    private var searchDisposable: Disposable? = null
    internal var authUrl: String? = null
    fun getLink(payment: Payment?): MutableLiveData<String> {
        if (searchDisposable != null) {
            searchDisposable!!.dispose()
        }
        val url = MutableLiveData<String>()
        apiService!!.getPaymentUrl(
            payment?.p,
            payment?.v_merchant_id,
            payment?.total,
            payment?.memo,
            payment?.cur
        )
            .subscribeOn(Schedulers.computation())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(object : Observer<String> {
                override fun onSubscribe(d: Disposable) {
                    searchDisposable = d
                }

                override fun onNext(s: String) {
                    url.value = s
                }

                override fun onError(e: Throwable) {
                    Log.e("Rep", e.message)
                }

                override fun onComplete() {
                    Log.e("TAG", "onComplete was reached")
                }
            }
            )
        return url
    }

    fun getLink(payment: WhiteLabel?): LiveData<LabelResponse?> {
        if (searchDisposable != null) {
            searchDisposable!!.dispose()
        }
        val gson = GsonBuilder().create()
        val payload = gson.toJson(payment)
        Log.e("repository", payload)
        val _url = MutableLiveData<LabelResponse?>()
        val url   = _url as LiveData<LabelResponse?>

        searchDisposable = apiService!!.getPaymentHeader(payload)
            .subscribeOn(Schedulers.computation())
            .observeOn(AndroidSchedulers.mainThread())
            .doOnError {
                Log.e("Rep", it.message)
            }
            .doOnComplete {
                Log.e("TAG", "onComplete was reached")

            }
            .subscribe{

                _url.postValue(it)
                if (it != null) {
                    Log.e("Rep", it.toString())
                }

            }

        return url
    }

}